### PR TITLE
TL/CUDA: allgather(v) linear alg

### PR DESCRIPTION
**Pull request by Sergei-Lebedev**
## What
Adding linear allgather algorithm in TL CUDA. Each rank broadcasts its source buffer to each other rank. Supported only on NVSwtich topology. Also added selection logic to allow user to choose what algorithm (ring or linear) to use.

### CODE REVIEW

**Comment by Sergei-Lebedev** [comment]
bot:retest

**Comment by bureddy** [comment]
BTW, are we testing different algorithms of collective somewhere(gtest, mtt) ?

File: src/components/tl/cuda/allgatherv/allgatherv_linear.c [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+
+        if (step % 2) {
+            remote_offset = ssize / 2;
+            local_offset = 0;
+        } else {
+            remote_offset = 0;
+            local_offset = ssize / 2;
+        }
+
+        if (step == 0) {
```
[the remark this answers is not in the document]
Comment (Sergei-Lebedev) [reply src/components/tl/cuda/allgatherv/allgatherv_linear.c:202]: added

File: src/components/tl/cuda/allgatherv/allgatherv_linear.c [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+            send_size = task->allgatherv_linear.get_count(task, trank);
+            frag_size = ucc_buffer_block_count(send_size, nfrags, step) * dt_size;
+            frag_offset = ucc_buffer_block_offset(send_size, nfrags, step) * dt_size;
```
[the remark this answers is not in the document]
Comment (Sergei-Lebedev) [reply src/components/tl/cuda/allgatherv/allgatherv_linear.c:205]: fixed

File: src/components/tl/cuda/allgatherv/allgatherv_linear.c [hunk 3]
```diff
@@ -0,0 +1,374 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2022.  ALL RIGHTS RESERVED.
```
[the remark this answers is not in the document]
Comment (Sergei-Lebedev) [reply src/components/tl/cuda/allgatherv/allgatherv_linear.c]: fixed

**Comment by Sergei-Lebedev** [comment]
> BTW, are we testing different algorithms of collective somewhere(gtest, mtt) ?

no so far, but in next PR I will add automatic alg selection based on topology, different algorithms will be triggered on different setups

File: src/components/tl/cuda/allgatherv/allgatherv_linear.c [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+{
+    ucc_tl_cuda_team_t *team  = TASK_TEAM(task);
+    return ucc_tl_cuda_shm_barrier_test(UCC_TL_TEAM_RANK(team), task->bar);
+}
+
+static inline size_t get_scratch_size(ucc_tl_cuda_team_t *team,
+                                      ucc_datatype_t dt)
+{
+    size_t dt_size = ucc_dt_size(dt);
+    ucc_rank_t tsize = UCC_TL_TEAM_SIZE(team);
```
Comment (vspetrov) [inline src/components/tl/cuda/allgatherv/allgatherv_linear.c:111]: align = (maybe run git-clang-format)

File: src/components/tl/cuda/allgatherv/allgatherv_linear.c [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+                                            local_offset + scratch_offset);
+                dbuf           = PTR_OFFSET(task->allgatherv_linear.rbuf,
+                                            rank_offset + frag_offset);
+
+                st = ecopy(dbuf, sbuf, frag_size, exec,
+                           &task->allgatherv_linear.exec_task[i]);
+                if (ucc_unlikely(st != UCC_OK)) {
+                    return st;
+                }
+
```
Comment (vspetrov) [inline src/components/tl/cuda/allgatherv/allgatherv_linear.c:265]: empty line

**Review by vspetrov — APPROVED** [review]
Plz add constructor to tl/cuda that initializes alg_info (e.g. https://github.com/openucx/ucc/blob/a7bda274b10f8adf5bb729f01da064f4e735fb23/src/components/tl/ucp/tl_ucp.c#L275) This is needed so that ucc_info -A printed available algorithms for TL/CUDA

File: src/components/tl/cuda/allgatherv/allgatherv_linear.c (hunk 4 above)
Comment (Sergei-Lebedev) [reply src/components/tl/cuda/allgatherv/allgatherv_linear.c]: done

**Comment by Sergei-Lebedev** [comment]
> Plz add constructor to tl/cuda that initializes alg_info (e.g.
> 
> https://github.com/openucx/ucc/blob/a7bda274b10f8adf5bb729f01da064f4e735fb23/src/components/tl/ucp/tl_ucp.c#L275
> 
> ) This is needed so that ucc_info -A printed available algorithms for TL/CUDA

added

**Review by bureddy — APPROVED** [review]

**Pull request merged**